rust/protocol/src/error.rs

```diff
@@ -11,82 +15,208 @@ use uuid::Uuid;
 
 use std::panic::UnwindSafe;
 
+#[cfg(doc)]
+pub use crate::{
+    address::ProtocolAddress,
+    curve::{PrivateKey, PublicKey},
+    group_cipher::{group_decrypt, group_encrypt},
+    protocol::{SenderKeyMessage, SignalMessage, CIPHERTEXT_MESSAGE_CURRENT_VERSION},
+    sealed_sender::{sealed_sender_decrypt, sealed_sender_multi_recipient_encrypt},
+    sender_keys::SenderKeyRecord,
+    session_cipher::{message_decrypt, message_encrypt},
+    state::{PreKeyId, PreKeyRecord, SignedPreKeyId, SignedPreKeyRecord},
+    storage::{IdentityKeyStore, PreKeyStore, SenderKeyStore, SessionStore, SignedPreKeyStore},
+};
+
+/// Return type for all fallible operations in this crate.
 pub type Result<T> = std::result::Result<T, SignalProtocolError>;
 
+/// Error states recognized by the Signal Protocol.
 #[derive(Debug, Display, Error)]
+#[ignore_extra_doc_attributes]
 pub enum SignalProtocolError {
     /// invalid argument: {0}
+    ///
+    /// Raised if an invalid argument is provided to any Signal API methods.
+    ///
+    /// Prefer to use lifetimes, static-sized slices, and dedicated wrapper structs in API
+    /// signatures to minimize the need to raise this error to FFI boundaries.
```

Comment:
Documentation should be client-facing, so comments about API design don't really fit.